public/app/index.ts

```diff
@@ -1,4 +1,5 @@
 import app from './app';
 
+app.initSentry();
```

Comment:
@dprokop -- can you take a look?  How close/different is this from the existing echo srv?  could we add a new interface to that?

Reply:
@ryantxu my thinking about Sentry reporter was to use EchoSrv as suggested on Slack @domasx2. 

[EchoSrv](https://github.com/grafana/grafana/blob/9e3def4214327dc1efd74fd68913c32f46c24524/packages/grafana-runtime/src/services/EchoSrv.ts#L16) attaches [default meta information to every event](https://github.com/grafana/grafana/blob/f48169633ce9a9387d468081140b1b1a6e2c5ad6/public/app/core/services/echo/Echo.ts#L70) that can be used to construct sentry events when flushing those to the backend, so the initSentry function that @domasx2 you have in this PR can be moved to [EchoBackend](https://github.com/grafana/grafana/blob/9e3def4214327dc1efd74fd68913c32f46c24524/packages/grafana-runtime/src/services/EchoSrv.ts#L52-L51) implementation for Sentry. 

@domasx2 if you want to read more about EchoSrv please take a look at the PR that introdced it https://github.com/grafana/grafana/pull/20365. For a custom async reporter you can have a look ar [MetaAnalyticsBackend](https://github.com/grafana/grafana-enterprise/blob/master/src/public/meta-analytics/MetaAnalyticsBackend.ts) from enterprise.

Not sure if Sentry's transport actually buffers the events before sending, but I think this would be useful for error reporting, especially for the ones that we catch, as those probably do not need to be reported immediately. 

@domasx2 let me know if there is anything unclear here or if me and @ryantxu  are missing something.

Reply:
Thanks @dprokop, I do have concerns:

* `initSentry` does more than configure transport - it wraps global functions, adds various listeners and so on to intercept errors. Also, this is best done as early as possible. Are you sure that an `EchoBackend` is the right place for this? I got the impression that it's only a transport layer

* I looked around, but don't see if `EchoSrv` will flush events if window is about to be closed. I think this is important for error reporting

* Delayed reporting is a concern for logging purposes imho, especially with regards to loki - we can't backdate a log timestamp. Events will appear significantly later than they happened. It's not possible to make it perfect, but would like to make a best effort
I guess we could log original timestamp as a separate field, but then it's not so convenient to investigate

* General unease about relying on system that is being monitored for the implementation of the monitoring :) In my experience error (and metrics!) reporting should be as decoupled as possible. I was even thinking we could configure webpack to spit out a separate entrypoint with Sentry setup that would preceed the main app

Anywyay, let me know what you think. If that's not convincing, I'll try to use echo backend and see how it goes

Reply:
> initSentry does more than configure transport - it wraps global functions, adds various listeners and so on to intercept errors. Also, this is best done as early as possible. Are you sure that an EchoBackend is the right place for this? I got the impression that it's only a transport layer

EchoSrv allows collecting events and report those to backends of interest on an interval (it basically calls flush on each registered backend). It's the Echo backend that is the transport. Echo is pretty much the first thing initialized in the app, so if Sentry was Echo backend then you get this without the need for `initSentry` function. Since backends are basically classes you could move the sentry config you are doing currently in the `initSentry` function to SentryBackend's constructor and achieve the same behavior.

An additional benefit of Echo is that if needed we can easily implement another Echo backend to consume errors and report to different endpoint/3rd party fairly easily as echo backend consume event types they are interested in (https://github.com/grafana/grafana/blob/f48169633ce9a9387d468081140b1b1a6e2c5ad6/public/app/core/services/echo/Echo.ts#L62). And we can expose `reportError` function, similar to [reportPerformance](https://github.com/grafana/grafana/blob/178bb1d3abbf45c62203cb1bff054bb474a9b4c2/public/app/core/services/echo/EchoSrv.ts#L4-L3) to allow imperative error collection which I think might be interesting for us as well.

> I looked around, but don't see if EchoSrv will flush events if window is about to be closed. I think this is important for error reporting

Yes, this is good point and something that we need to add to Echo.

> Delayed reporting is a concern for logging purposes imho, especially with regards to loki - we can't backdate a log timestamp. Events will appear significantly later than they happened. It's not possible to make it perfect, but would like to make a best effort
I guess we could must log original timestamp as a separate field, but then it's not so convenient to investigate

[Each Echo event has timestamp attached to it](https://github.com/grafana/grafana/blob/f48169633ce9a9387d468081140b1b1a6e2c5ad6/public/app/core/services/echo/Echo.ts#L70). I'm not familiar with how Loki treat logs, but is it the timestamp is added to the log when it's reported? Can't you use the timestamp passed with the log as a source for log timing?


> General unease about relying on system that is being monitored for the implementation of the monitoring :) In my experience error (and metrics!) reporting should be as decoupled as possible. I was even thinking we could configure webpack to spit out a separate entrypoint with Sentry setup that would preceed the main app

I agree and I think this sounds like a good idea going forward.

Reply:
> Since backends are basically classes you could move the sentry config you are doing currently in the initSentry function to SentryBackend's constructor and achieve the same behavior. An additional benefit of Echo is that if needed we can easily implement another Echo backend to consume errors and report to different endpoint/3rd party fairly easily as echo backend consume event types they are interested in.

I get the benefits of a decoupled event ingestion & procesing. 
What would we do if Grafana user wants to report to Sentry backend? In current version we don't configure any custom transport, just let Sentry do it's thing. Should we then still have Sentry dump to `EchoSrv` and somehow marry Sentry internal transport with `SentryEchoBackend`?

> And we can expose reportError function, similar to reportPerformance to allow imperative error collection which I think might be interesting for us as well.

Sentry exposes `captureError` for this, and we must use it rather than pushing to `EchoSrv` direclty to have Sentry gather metadata & instantiate the event.

> I'm not familiar with how Loki treat logs, but is it the timestamp is added to the log when it's reported? Can't you use the timestamp passed with the log as a source for log timing?

My understanding is that Loki currently requires timestamps to be in-order. So it only makes sense to log with timestamp generated at the time the log is ingested in backend, to be in order with other backend logs. There's no way to sort out order for events coming asynchrnously from multiple browsers anyway. I'll try to clarify this with Loki guys though

> I agree and I think this sounds like a good idea going forward.
So all in all, move Sentry init to separate entrypooint? With or without EchoSrv, what's the conclusion? :)

Reply:
Regarding logs I had a chat with Ed:
In summary, surrently logs have to be in-order. This restrition might be lifted in next 3-6 months.  A workaround would be to have log stream per browser session, but is a bad use case for Loki due to high cardinality.
So currenly sounds like best option is to use the timestamp of when the log is output to stdout at backend, so it's in order with other logs for the same stream

Reply:
@domasx2 ok, I think I  see the complete constraints around this work. Correct me if I'm wrong but Sentry will basically send the captured error as soon as it occured, right?

Going forward I think we should do the following:

1. Move sentry configuration (basically initSentry function) into Echo backend. 
2. This backend, since we will be relying on Sentry's transport, would implement `flush` as a noop function and encapsulate the Sentry's `captureError` function within Sentry's backend `addEvent` implementation and expose it indirectly via `reportError` function, similar to what we do with `reportPerformance`. This will enable us in the future to have a different/additional backend for error reports 
3. I wonder if we could hijack automatic Sentry's error capturing so that it also adds events to Sentry's Echo backend if we would like to report the errors to some other storage independently from Sentry's transport. **I don't think we need to do this now though**!

> So all in all, move Sentry init to separate webpack entrypoint? With or without EchoSrv, what's the conclusion? :)

I think you can try this out, taking the initEchoSrv out from the main app and initializing this in a separate chunk.

Does ^^ make sense? I'm pushing towards using Echo (even though not utilizing it's flush behavior now), so that all captured "events" go through the same pipeline.

Reply:
Ok, if we can send errors immediately, I'm onboard with using EchoSrv :) 

If I understand correctly, you suggest that `SentryBackend.addEvent` would call `Sentry.captureError`. This has a flaw however - only errors reported via `addEvent` would make it to `EchoSrv`, but not erorrs captured by Sentry itself through the various hooks it sets up on promises, `window.onerror`, etc.

Then I propose to invert it: I'll override Sentry transport to dump all events into `EchoSrv`, and make `SentryBackend` encapsulate `SentryTransport` to do the http sending. This way all errors will go through `EchoSrv`. Makes sense?

Reply:
- [ ] send all events to `EchoSrv`, attach backends to consume them. Can flush immediately